File: src/kicad_xyce_plugin/simulation_parameters/option_parameters.py
from __future__ import annotations

from dataclasses import dataclass
from dataclasses import field

from netlist_parser import NetlistTopology


def _parse_option_tokens(tokens: list[str]) -> dict[str, str]:
    # parse a series of option tokens into a normalized dictionary
    options: dict[str, str] = {}
    for token in tokens:
        # skip empty tokens produced by extra whitespace
        if not token:
            continue
        # split key/value pairs and normalize keys to uppercase
        if "=" in token:
            key, value = token.split("=", 1)
            options[key.upper()] = value
            continue
        # support flag-style options without an explicit value
        options[token.upper()] = ""
    return options


@dataclass(frozen=True)
class OptionParameters:
    # top-level device options applied across simulations
    device: dict[str, str] = field(default_factory=dict)
    # transient integration control parameters
    timeint: dict[str, str] = field(default_factory=dict)
    # generic nonlinear solver parameters
    nonlin: dict[str, str] = field(default_factory=dict)
    # generic linear solver parameters
    linsol: dict[str, str] = field(default_factory=dict)

    @classmethod
    def from_xyce_directives(cls, directives: list[str]) -> "OptionParameters":
        # init option groups
        device: dict[str, str] = {}
        timeint: dict[str, str] = {}
        nonlin: dict[str, str] = {}
        linsol: dict[str, str] = {}
        # parse each directive looking for supported option packages
        for directive in directives:
            # break directive into tokens
            tokens = directive.split()
            # skip empty directives
            if not tokens:
                continue
            # handle only .OPTIONS directives
            if tokens[0].upper() != ".OPTIONS" or len(tokens) <= 1:
                continue
            # normalize the package name
            package = tokens[1].upper()
            if package == "DEVICE":
                device = _parse_option_tokens(tokens[2:])
                continue
            if package == "TIMEINT":
                timeint = _parse_option_tokens(tokens[2:])
                continue
            if package == "NONLIN":
                nonlin = _parse_option_tokens(tokens[2:])
                continue
            if package == "LINSOL":
                linsol = _parse_option_tokens(tokens[2:])
                continue
        return cls(device=device, timeint=timeint, nonlin=nonlin, linsol=linsol)

    def to_xyce_directives(self, topology: NetlistTopology | None = None) -> list[str]:
        # serialize configured option blocks in a deterministic order
        directives: list[str] = []
        if self.device:
            directives.append(".OPTIONS DEVICE " + " ".join(f"{k}={v}" if v else k for k, v in self.device.items()))
        if self.timeint:
            directives.append(".OPTIONS TIMEINT " + " ".join(f"{k}={v}" if v else k for k, v in self.timeint.items()))
        if self.nonlin:
            directives.append(".OPTIONS NONLIN " + " ".join(f"{k}={v}" if v else k for k, v in self.nonlin.items()))
        if self.linsol:
            directives.append(".OPTIONS LINSOL " + " ".join(f"{k}={v}" if v else k for k, v in self.linsol.items()))
        return directives

File: src/simulation/option_parameters.cpp
#include <cctype>
#include <string>
#include <vector>

#include "../core/util.h"
#include "option_parameters.h"

struct NetlistTopology;

// parse a series of option tokens into a normalized map
static std::map<std::string, std::string> parse_option_tokens(const std::vector<std::string>& tokens) {
    // parse a series of option tokens into a normalized dictionary
    std::map<std::string, std::string> options;
    for (const auto& token : tokens) {
        // skip empty tokens produced by extra whitespace
        if (token.empty()) {
            continue;
        }
        // split key/value pairs and normalize keys to uppercase
        const auto eq_pos = token.find('=');
        if (eq_pos != std::string::npos) {
            const std::string key = to_upper(token.substr(0, eq_pos));
            const std::string val = token.substr(eq_pos + 1);
            options[key] = val;
            continue;
        }
        // support flag-style options without an explicit value
        options[to_upper(token)] = "";
    }
    return options;
}

OptionParameters::OptionParameters(std::map<std::string, std::string> device, std::map<std::string, std::string> timeint, std::map<std::string, std::string> nonlin, std::map<std::string, std::string> linsol, std::map<std::string, std::string> fft, std::map<std::string, std::string> diagnostic) :
    device(std::move(device)), timeint(std::move(timeint)), nonlin(std::move(nonlin)), linsol(std::move(linsol)), fft(std::move(fft)), diagnostic(std::move(diagnostic)) {}

OptionParameters OptionParameters::from_xyce_directives(const std::vector<std::string>& directives) {
    // init option groups
    std::map<std::string, std::string> device;
    std::map<std::string, std::string> timeint;
    std::map<std::string, std::string> nonlin;
    std::map<std::string, std::string> linsol;
    std::map<std::string, std::string> fft;
    std::map<std::string, std::string> diagnostic;

    // parse each directive looking for supported option packages
    for (const auto& directive : directives) {
        // break directive into tokens
        std::vector<std::string> tokens;
        std::string current;
        for (const char ch : directive) {
            if (std::isspace(static_cast<unsigned char>(ch))) {
                if (!current.empty()) {
                    tokens.push_back(current);
                    current.clear();
                }
                continue;
            }
            current += ch;
        }
        if (!current.empty()) {
            tokens.push_back(current);
        }

        // skip empty directives
        if (tokens.empty()) {
            continue;
        }

        // handle only .OPTIONS directives
        if (to_upper(tokens[0]) != ".OPTIONS" || tokens.size() <= 1) {
            continue;
        }

        // normalize the package name
        const std::string package = to_upper(tokens[1]);

        if (package == "DEVICE") {
            device = parse_option_tokens(std::vector<std::string>(tokens.begin() + 2, tokens.end()));
            continue;
        }
        if (package == "TIMEINT") {
            timeint = parse_option_tokens(std::vector<std::string>(tokens.begin() + 2, tokens.end()));
            continue;
        }
        if (package == "NONLIN") {
            nonlin = parse_option_tokens(std::vector<std::string>(tokens.begin() + 2, tokens.end()));
            continue;
        }
        if (package == "LINSOL") {
            linsol = parse_option_tokens(std::vector<std::string>(tokens.begin() + 2, tokens.end()));
            continue;
        }
        if (package == "FFT") {
            fft = parse_option_tokens(std::vector<std::string>(tokens.begin() + 2, tokens.end()));
            continue;
        }
        if (package == "DIAGNOSTIC") {
            diagnostic = parse_option_tokens(std::vector<std::string>(tokens.begin() + 2, tokens.end()));
            continue;
        }
    }

    return OptionParameters(device, timeint, nonlin, linsol, fft, diagnostic);
}

std::vector<std::string> OptionParameters::to_xyce_directives(const NetlistTopology& topology) const {
    (void)topology;
    // serialize configured option blocks in a deterministic order
    std::vector<std::string> directives;

    // build helper lambda to format options
    auto format_options = [](const std::map<std::string, std::string>& options) -> std::string {
        std::string result;
        for (const auto& [key, value] : options) {
            if (!result.empty()) {
                result += " ";
            }
            if (!value.empty()) {
                result += key + "=" + value;
            }
            else {
                result += key;
            }
        }
        return result;
    };

    if (!device.empty()) {
        directives.push_back(".OPTIONS DEVICE " + format_options(device));
    }
    if (!timeint.empty()) {
        directives.push_back(".OPTIONS TIMEINT " + format_options(timeint));
    }
    if (!nonlin.empty()) {
        directives.push_back(".OPTIONS NONLIN " + format_options(nonlin));
    }
    if (!linsol.empty()) {
        directives.push_back(".OPTIONS LINSOL " + format_options(linsol));
    }
    if (!fft.empty()) {
        directives.push_back(".OPTIONS FFT " + format_options(fft));
    }
    if (!diagnostic.empty()) {
        directives.push_back(".OPTIONS DIAGNOSTIC " + format_options(diagnostic));
    }

    return directives;
}

bool OptionParameters::operator==(const OptionParameters& other) const {
    // compare all fields for equality
    return device == other.device && timeint == other.timeint && nonlin == other.nonlin && linsol == other.linsol && fft == other.fft && diagnostic == other.diagnostic;
}
